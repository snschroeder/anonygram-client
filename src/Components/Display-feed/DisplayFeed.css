.display-feed {
    position: relative;
    margin-bottom: 7em;
    width: 90%;
    margin: 0 auto;
}

.img-container {
    margin: 0;
    padding: 0;
    display: grid;
    grid-row-gap: 10px;
    grid-column-gap: 14px;
}

@media screen and (min-width: 50em) {
    .img-container {
        grid-template-columns: repeat(4, 1fr);
    }
    .img-container li {
        grid-column: span 2;
        grid-row: span 2;
    }
}

@media screen and (min-width: 768px) {
    .display-feed {
        margin-bottom: 3em;
    }
}

@media screen and (min-width: 80em) {
    .display-feed {	
        padding-bottom: 5em;	
    }
}

@media screen and (min-width: 1120px) {
    .img-container {
        max-width: 990px;
        margin: 0 calc((100vw - 990px)/2);
    }
}
