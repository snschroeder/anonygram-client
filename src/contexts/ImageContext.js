import React, { useContext } from 'react';
export const useImageContext = () => useContext(ImageContext);
const defaultValues = {
  //vars in ABC order
  alert: '',
  debounce: null,
  error: null,
  images: [],
  morePagesAvail: null,
  newContentLoaded: null, 
  page: null,
  sort: ['new', 'top'],
  user: null,
  userLocation: {},
  //funcs in ABC order
  clearAlert: () => {},
  clearError: () => {},
  handleDelete: () => {},
  incrementUpvotes: () => {},
  resetState: () => {},
  setAlert: () => {},
  setDebounce: () => {},
  setError: () => {},
  setMorePagesAvail: () => {},
  setNewContentLoaded: () => {},
  setPage: () => {},
  setView: () => {},
}
const ImageContext = React.createContext(defaultValues)

export default ImageContext;
