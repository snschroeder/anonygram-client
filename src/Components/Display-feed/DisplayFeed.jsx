import React, { useState, useEffect, useRef } from 'react';
import DisplayItem from './Display-item/DisplayItem';
import SubmissionForm from '../SubmissionForm/SubmissionForm';
import { useImageContext } from '../../contexts/ImageContext';

import './DisplayFeed.css';

export default function DisplayFeed(props) {
    const [pageOffset, setPageOffset] = useState(1000);
    const [bottom, setBottom] = useState(2000);

    const context = useImageContext();
    const dpFeedRef = useRef(null);

    const { images, page, setPage, morePagesAvail, debounce, handleDelete, incrementUpvotes } = context;

    const { userLocation, newContentLoaded, updateNewContent, ...rest } = props

    useEffect(() => {
        props.setView('feed');
        setPageOffset(window.pageYOffset);
        setBottom(dpFeedRef.current.getBoundingClientRect().bottom);
        const handleScroll = () => {
            setPageOffset(window.pageYOffset);
        }
        window.addEventListener('scroll', handleScroll);
        return (() => {
            window.removeEventListener('scroll', handleScroll);
        })
    }, []);

    useEffect(() => {
        if (dpFeedRef.current) {
            setBottom(dpFeedRef.current.getBoundingClientRect().bottom);
        }

        if (pageOffset >= bottom && morePagesAvail && !debounce) {
            let pageClone = page + 1;
            setPage(pageClone);
            setPageOffset(window.pageYOffset);
        };

    }, [bottom, pageOffset, morePagesAvail, page, setPage, debounce]);

    const generateJSX = () => {

        if (!images) {
            return null;
        }
        return (
            <>
                <ul className="img-container" ref={dpFeedRef}>
                    {images.map(imgObj => (
                        <DisplayItem
                            imgAddress={imgObj.image_url}
                            imgCaption={imgObj.image_text}
                            upvotes={imgObj.karma_total}
                            id={imgObj.id}
                            userId={imgObj.user_id}
                            incrementUpvotes={incrementUpvotes}
                            handleDelete={handleDelete}
                            key={imgObj.id}
                        />
                    ))}
                </ul>
                <SubmissionForm
                  routeProps={rest}
                  userLocation={userLocation}
                  newContentLoaded={newContentLoaded}
                  updateNewContent={updateNewContent}
                  parent='DisplayFeed'
                />
            </>
        )
    }

    return (
        <section className="display-feed">
            {generateJSX()}
        </section>
    )
}

DisplayFeed.defaultProps = {
    setView: () => {}
}