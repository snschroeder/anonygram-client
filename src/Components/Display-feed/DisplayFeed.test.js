import React from 'react';
import ReactDOM from 'react-dom';
import DisplayFeed from './DisplayFeed';
import * as ImageContext from '../../contexts/ImageContext';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('DisplayFeed component', () => {
  const contextValues = {
    images: [
      {
        id: 20,
        image_url:
          'https://lh3.googleusercontent.com/p/AF1QipMJcAXkTwnOj_TWYbMRwygy8ktducIF2kgOqSdW=s1600-w500',
        image_text: null,
        karma_total: 0,
        latitude: '29.65468429999999',
        longitude: '-82.4158946',
        create_timestamp: '2019-12-10T00:40:55.725Z',
      },
      {
        id: 21,
        image_url:
          'https://lh3.googleusercontent.com/p/AF1QipMJcAXkTwnOj_TWYbMRwygy8ktducIF2kgOqSdW=s1600-w500',
        image_text: 'some optional text',
        karma_total: 0,
        latitude: '29.65468429999999',
        longitude: '-82.4158946',
        create_timestamp: '2019-12-11T00:40:55.725Z',
      },
    ],
    incrementUpvotes: () => contextValues.images[0].karma_total += 1,
  };
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<DisplayFeed />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders elements by default', () => {
    const wrapper = shallow(<DisplayFeed />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('renders elements given ImageContext', async () => {
    jest
      .spyOn(ImageContext, 'useImageContext')
      .mockImplementation(() => contextValues);
    const wrapper = await shallow(<DisplayFeed />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('clicking upvote updates prop value as expected', () => {
    const context = {...contextValues}
    jest
      .spyOn(ImageContext, 'useImageContext')
      .mockImplementation(() => contextValues);
      const wrapper = shallow(<DisplayFeed />, { context })
      // "click" it twice
      contextValues.images[0].karma_total = wrapper.find('DisplayItem').at(0).props().incrementUpvotes()
      contextValues.images[0].karma_total = wrapper.find('DisplayItem').at(0).props().incrementUpvotes()
      wrapper.setContext(contextValues)
      expect(toJson(wrapper)).toMatchSnapshot();
  })
});
