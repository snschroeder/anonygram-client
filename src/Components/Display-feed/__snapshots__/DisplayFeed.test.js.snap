// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayFeed component clicking upvote updates prop value as expected 1`] = `
<section
  className="display-feed"
>
  <ul
    className="img-container"
  >
    <DisplayItem
      id={20}
      imgAddress="https://lh3.googleusercontent.com/p/AF1QipMJcAXkTwnOj_TWYbMRwygy8ktducIF2kgOqSdW=s1600-w500"
      imgCaption={null}
      incrementUpvotes={[Function]}
      key="20"
      upvotes={2}
    />
    <DisplayItem
      id={21}
      imgAddress="https://lh3.googleusercontent.com/p/AF1QipMJcAXkTwnOj_TWYbMRwygy8ktducIF2kgOqSdW=s1600-w500"
      imgCaption="some optional text"
      incrementUpvotes={[Function]}
      key="21"
      upvotes={0}
    />
  </ul>
  <SubmissionForm
    parent="DisplayFeed"
    routeProps={
      Object {
        "setView": [Function],
      }
    }
  />
</section>
`;

exports[`DisplayFeed component renders elements by default 1`] = `
<section
  className="display-feed"
>
  <ul
    className="img-container"
  />
  <SubmissionForm
    parent="DisplayFeed"
    routeProps={
      Object {
        "setView": [Function],
      }
    }
  />
</section>
`;

exports[`DisplayFeed component renders elements given ImageContext 1`] = `
<section
  className="display-feed"
>
  <ul
    className="img-container"
  >
    <DisplayItem
      id={20}
      imgAddress="https://lh3.googleusercontent.com/p/AF1QipMJcAXkTwnOj_TWYbMRwygy8ktducIF2kgOqSdW=s1600-w500"
      imgCaption={null}
      incrementUpvotes={[Function]}
      key="20"
      upvotes={0}
    />
    <DisplayItem
      id={21}
      imgAddress="https://lh3.googleusercontent.com/p/AF1QipMJcAXkTwnOj_TWYbMRwygy8ktducIF2kgOqSdW=s1600-w500"
      imgCaption="some optional text"
      incrementUpvotes={[Function]}
      key="21"
      upvotes={0}
    />
  </ul>
  <SubmissionForm
    parent="DisplayFeed"
    routeProps={
      Object {
        "setView": [Function],
      }
    }
  />
</section>
`;
